pyzx/graph/multigraph.py

```diff
@@ -0,0 +1,380 @@
+
+    def remove_vertex(self, vertex):
+        self.remove_vertices([vertex])
+
+    def remove_edges(self, edges):
+        for e in edges:
+            self.remove_edge(e)
+
+    def remove_edge(self, edge):
+        s,t,ty = edge
+        e = self.graph[s][t]
+        if ty == EdgeType.SIMPLE: e.remove(s=1)
+        else: e.remove(h=1)
+
+        if e.s == 0 and e.h == 0:
+            del self.graph[s][t]
+            del self.graph[t][s]
+
+        self.nedges -= 1
+
+    def num_vertices(self):
+        return len(self.graph)
+
+    def num_edges(self):
+        return self.nedges
+        #return len(self.edge_set())
+
+    def vertices(self):
+        return self.graph.keys()
+
+    def vertices_in_range(self, start, end):
+        """Returns all vertices with index between start and end
+        that only have neighbours whose indices are between start and end"""
+        for v in self.graph.keys():
+            if not start<v<end: continue
+            if all(start<v2<end for v2 in self.graph[v]):
+                yield v
+
+    def edges(self, s=None, t=None):
+        if s == None:
+            for v0,adj in self.graph.items():
+                for v1, e in adj.items():
+                    if v1 > v0:
+                        for _ in range(e.s): yield (v0, v1, EdgeType.SIMPLE)
+                        for _ in range(e.h): yield (v0, v1, EdgeType.HADAMARD)
+        elif t != None:
+            s, t = (s, t) if s < t else (t, s)
+            e = self.graph[s][t]
+            for _ in range(e.s): yield (s, t, EdgeType.SIMPLE)
+            for _ in range(e.h): yield (s, t, EdgeType.HADAMARD)
+
+    # def edges_in_range(self, start, end, safe=False):
+    #    """like self.edges, but only returns edges that belong to vertices
+    #    that are only directly connected to other vertices with
+    #    index between start and end.
+    #    If safe=True then it also checks that every neighbour is only connected to vertices with the right index"""
+    #    if not safe:
+    #        for v0,adj in self.graph.items():
+    #            if not (start<v0<end): continue
+    #            #verify that all neighbours are in range
+    #            if all(start<v1<end for v1 in adj):
+    #                for v1 in adj:
+    #                    if v1 > v0: yield (v0,v1)
+    #    else:
+    #        for v0,adj in self.graph.items():
+    #            if not (start<v0<end): continue
+    #            #verify that all neighbours are in range, and that each neighbour
+    #            # is only connected to vertices that are also in range
+    #            if all(start<v1<end for v1 in adj) and all(all(start<v2<end for v2 in self.graph[v1]) for v1 in adj):
+    #                for v1 in adj:
+    #                    if v1 > v0:
+    #                        yield (v0,v1)
+
+    def edge(self, s, t):
+        return (s,t) if s < t else (t,s)
+    def edge_set(self):
+        return set(self.edges())
+    def edge_st(self, edge):
```

Comment:
`edge_set` and `edge_st` are dangerously close to "minor typo leads to frustrating-to-find bug" territory, though fortunately they have different signatures.